### PR TITLE
SPF query UI - FAD-4221

### PR DESCRIPTION
**Pull request by jasonrhodes**

### CODE REVIEW

File: src/pages/spf/Query.js [hunk 1]
```diff
@@ -16,45 +17,46 @@ export default class Query extends Component {
...
-    this.setState({domainError: this.validateDomain(domain)});
-
-    if (!this.state.domainError) {
-      // go to `/spf/results/${this.state.domain}`
-    }
+  componentDidMount() {
+    this.props.router.setRouteLeaveHook(this.props.route, () => {
+      const validDomain = this.validateDomain(this.state.domain);
+      this.setState({domainError: !validDomain});
+      return validDomain;
```
Comment (jonambas) [inline src/pages/spf/Query.js]: Alternative:
- Add `onClick` to submit button
- Run validation
- `this.props.router.push('/spf/results/${this.state.domain}');`

File: src/pages/spf/components/ChildRecord.js [hunk 2]
```diff
@@ -12,10 +12,18 @@ export default class ChildRecord extends Component {
...
+    // A component cannot update its own props
+    //
+    // const { record } = this.props;
+    //
+    // record.displayType = record.type;
+    //
+    // if (record.type === 'mx' && this.hasChildren()) {
+    //   record.children.forEach((child) => {
+    //     child.type = 'mxChild';
+    //     child.displayType = 'mx';
```
Comment (jonambas) [inline src/pages/spf/components/ChildRecord.js]: So a component cannot change its own props in React, I don't think I fully understand what's going on here and in the render function though.

So things like `this.props.record.displayType = this.props.record.type;` won't work

Comment (jonambas) [inline src/pages/spf/components/ChildRecord.js]: I think the cleaner solution is to reorganize mx children after the api call in Results.js, or in a reducer if we do Redux

File: src/routes.js [hunk 3]
```diff
@@ -13,6 +15,10 @@ export default (
       <Route path='dkim' component={DKIMHome} />
       <Route path='dkim/results/:email' component={DKIMResults} />
       <Route path='dkim/results/:email/:detailId' component={DKIMDetail} />
+
+      <Route path='spf-inspector' component={SPFQuery} />
```
Comment (jasonrhodes) [inline src/routes.js]: I'm not crazy about this change but I think that's related to other spf things, so we should revisit this after our kickoff on Monday.

File: src/pages/spf/components/SPFTree.scss [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    left: 0;
+    top: 0;
+    bottom: 0;
+    width: 3px;
+  }
+
+  &__children {
+    position: relative;
+
+    // Yo this selector is crazy
```
Comment (jasonrhodes) [inline src/pages/spf/components/SPFTree.scss]: ❤️ 💟 😍

File: src/pages/spf/components/ResultsErrors.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      </div>
+
+      { errors.map((error, idx) => (
+        <div key={`e-${idx}`} className="body__panel">
+          <i className="fa fa-exclamation-circle"></i>
+          <span>{error.message}</span>
+        </div>
+      )) }
+
+      { errors.map((error, idx) => (
```
Comment (jonambas) [inline src/pages/spf/components/ResultsErrors.js]: Oops, this should be warnings.map

**Comment by jasonrhodes** [comment]
So I've been playing with this a bit and I think our best bet is to put every node in the tree into an "expanded" state by default, and not allow any collapsing in this ticket at all, the collapse button can do nothing for right now or we can hide it. I've got a good redux solution that should solve this pretty well and also give us a way to deep link later.

I've got a couple other ideas about the recursive components, too, that I can show in person or draw up a quick and dirty PR against this one so we can talk them out on Monday. :)

**Comment by jonambas** [comment]
@jasonrhodes I think we can allow collapsing in this ticket - maybe not collapse/expand ALL. But collapsing works right now

File: src/routes.js (hunk 3 above)
Comment (jasonrhodes) [inline src/routes.js]: OK so back to this, what about putting it at `/spf/inspector` or `/spf/validator` and then we can put together a simple `/spf` landing page that describes SPF and links to the inspector and eventually the wizard, once both are done?

Comment (jasonrhodes) [inline src/routes.js]: While we're at it we could move `/dkim` to `/dkim/validator` and just redirect that for now too.

**Pull request merged**